db/src/tx_observer.rs

```diff
@@ -23,34 +24,28 @@ use types::{
 };
 
 pub struct TxObserver {
-    notify_fn: Arc<Option<Box<Fn(String, Vec<TxReport>) + Send + Sync>>>,
+    notify_fn: Arc<Box<Fn(&String, &Vec<Arc<TxReport>>) + Send + Sync>>,
     attributes: AttributeSet,
 }
 
 impl TxObserver {
-    pub fn new<F>(attributes: AttributeSet, notify_fn: F) -> TxObserver where F: Fn(String, Vec<TxReport>) + 'static + Send + Sync {
+    pub fn new<F>(attributes: AttributeSet, notify_fn: F) -> TxObserver where F: Fn(&String, &Vec<Arc<TxReport>>) + 'static + Send + Sync {
         TxObserver {
-            notify_fn: Arc::new(Some(Box::new(notify_fn))),
+            notify_fn: Arc::new(Box::new(notify_fn)),
             attributes,
         }
     }
 
-    pub fn applicable_reports(&self, reports: &Vec<TxReport>) -> Vec<TxReport> {
+    pub fn applicable_reports(&self, reports: &Vec<Arc<TxReport>>) -> Vec<Arc<TxReport>> {
```

Comment:
I think you can make this

```
pub fn applicable_reports(&self, reports: &Vec<TxReport>) -> Vec<&TxReport> {
```

— the notifying code can take reports by ownership, and it always invokes the observers on the same thread and within the lifetime of the reports, so we can avoid using `Arc` at all, I think.

The main reasons one would want to use an `Arc` here is if:

1. The reports need to be held by two threads simultaneously. That isn't the case here, as far as I can tell, but maybe I'm wrong.
2. You expect observers to routinely want to keep whole reports for longer than the duration of the observer callback. I don't feel this is likely, but perhaps Toodle does?